### ISSUE TITLE
Support for embeddable objects

### ISSUE THREAD

**Issue by tlvenn** [neumino/thinky#86]
Hi Neumino,

I am playing with thinky and when it comes to relationship, the assumption is to leverage the join feature of rethinkdb which is great. However in some scenarios, it's preferable to embed rather than to join.

To some degrees, thinky might already support this with nested objects but it forces you to inline all your embeddables and I am not so sure it's ideal to support collection of elements.

Maybe if we could declare our embeddables with thinky.createEmbeddable and then reference them in the schema where they are embedded.

Now i might be biased by my prior JPA experience and maybe this whole thing is unnecessary when it comes to json documents...

What do you think ?

**Comment by cur3n4** [comment]
+1

I think the ability to define virtual fields which operate just on the nested model properties would be great. e.g:

```javascript
var User = thinky.createModel("User", {
  id: String,
  contact: {
    email: String,
    phone: String,
    fullContact: type.virtual().default(function() {
        return this.email+" "+this.phone;
    }
  }
});
```

**Comment by neumino** [comment]
@cur3n4 -- you can already do that.

**Comment by cur3n4** [comment]
@neumino Indeed! However it doesn't seem to work for arrays of nested objects. I get a TypeError: Cannot read property '0' of undefined.

Ideally defining methods on the nested model (not only virtual) would be very useful.

**Comment by neumino** [comment]
@cur3n4 -- can you open a new issue with the full backtrace and a snippet to reproduce this error? Thanks

**Comment by cur3n4** [comment]
@neumino Done: [#375](https://github.com/neumino/thinky/issues/375)

**Comment by cur3n4** [comment]
From my point of view, the main thing I am missing about nested models is the ability to define methods (not virtual properties). Something like this:

```javascript
var User = thinky.createModel("User", {
  id: String,
  contact: {
    email: String,
    someMethod: type.method().default(function(someParameter1, someParameter2) {
        // More complex logic than this...
        return this.email == someParameter1 || this.email == someParameter2;
    }
  }
});
```
